lib/compat.php

```diff
@@ -24,18 +24,90 @@
 	 * @return WP_Block_Type|false The registered block type on success, or false on failure.
 	 */
 	function register_block_type_from_metadata( $path, $args = array() ) {
-		$file = trailingslashit( $path ) . 'block.json';
-		if ( ! file_exists( $file ) ) {
+		$metadata_file = trailingslashit( $path ) . 'block.json';
+		if ( ! file_exists( $metadata_file ) ) {
 			return false;
 		}
 
-		$metadata = json_decode( file_get_contents( $file ), true );
-		if ( ! is_array( $metadata ) ) {
+		$metadata = json_decode( file_get_contents( $metadata_file ), true );
+		if ( ! is_array( $metadata ) || ! $metadata['name'] ) {
 			return false;
 		}
 
+		$block_dir           = dirname( $metadata_file );
+		$block_name          = $metadata['name'];
+		$asset_handle_prefix = str_replace( '/', '-', $block_name );
+
+		if ( isset( $metadata['editorScript'] ) ) {
+			$editor_script            = $metadata['editorScript'];
+			$editor_script_handle     = "$asset_handle_prefix-editor-script";
+			$editor_script_asset_path = realpath(
+				$block_dir . '/' . substr_replace( $editor_script, '.asset.php', -3 )
+			);
+			if ( ! file_exists( $editor_script_asset_path ) ) {
+				throw new Error(
+					"The asset file for the \"editorScript\" defined in \"$block_name\" block definition is missing."
+				);
+			}
+			$editor_script_asset  = require( $editor_script_asset_path );
+			wp_register_script(
+				$editor_script_handle,
+				plugins_url( $editor_script, $metadata_file ),
+				$editor_script_asset['dependencies'],
+				$editor_script_asset['version']
+			);
+			unset( $metadata['editorScript'] );
+			$metadata['editor_script'] = $editor_script_handle;
```

Comment:
Are you proposing here to discourage the camelCase?

Related: https://github.com/WordPress/gutenberg/pull/13693#issuecomment-612072560

We should probably want to be consistent? Thinking specifically the new `providesContext` and whether it should be converted similarly (separate from this pull request).

Reply:
As you pointed out in the related comment, there isn't much of the choice if we want to wire it with how `register_block_type` works today:

https://github.com/WordPress/wordpress-develop/blob/eb79d7f1ee0f59d1035260895b134f134a0ac703/src/wp-includes/script-loader.php#L2494-L2521

There are 4 fields handled and they all use snake_case:
- `style`
- `script`
- `editor_style`
- `editor_script`

In this PR I follow RFC and use camelCase in `block.json` to provide paths to files. `register_block_type_from_metadata` either overrides those values (`script` and `style`) with an asset handle or unsets paths (`editor_script` and `editor_style`) and sets asset handles using snake_case notation as the key.

Should I maybe make it more explicit and use `$settings` object as return value and explicitly map only allowed values?

Reply:
How about this explicit mapping between `$metadata` and `$settings`: https://github.com/WordPress/gutenberg/pull/22519/commits/139a1805c65ef2b75a3c0bbde4779f1fe4f9afba?

Reply:
If we get to the point where all blocks are registered on the server from `block.json`, and then bootstrapped into the editor using those server-registered properties, do we have to worry about mapping those snake_case properties back to camelCase for the editor?

Specifically talking about `unstable__bootstrapServerSideBlockDefinitions`

Reply:
I hope we move away soon from this unstable helper to using REST API endpoint for blocks (preloaded on the server) that has its own specific set of coding guidelines. I anticipate some sort of mapping is going to be necessary on the common entities level in the data layer, so I wouldn’t worry that much about it.